### ISSUE TITLE
model_dump uses parent class serialization

### ISSUE THREAD

**Issue by mirco-gnuva-unimi** [pydantic/pydantic#7963]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Good morning,
I facing a problem using the model_dump() method; I'm not sure if it is an issue or a design decision.
The model_dump() method seams to serialize attributes using theirs parent type instead of theirs direct one.
Thank you for you time.

### Example Code

```Python
class A(BaseModel):
    attr1: str
    attr2: int


class B(A):
    attr3: str


class C(BaseModel):
    attr: A


b = B(attr1='Test',
      attr2=5,
      attr3='Subclass')


c = C(attr=b)

dump = c.model_dump()

print(dump)


# Gives: {'attr': {'attr1': 'Test', 'attr2': 5}}
# I would expect: {'attr': {'attr1': 'Test', 'attr2': 5, 'attr3': 'Subclass'}}
# The expected behaviour can be obtained declaring attr of type B, or Union[A,B]
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
pydantic-core version: 2.10.1
pydantic-core build: profile=release pgo=false
python version: 3.11.4 (main, Jul 18 2023, 14:10:40) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-14.1-arm64-arm-64bit
             related packages: email-validator-1.1.2 typing_extensions-4.8.0 fastapi-0.104.0 pydantic-settings-2.0.3
```

**Comment by samuelcolvin** [comment]
This is a question, not a bug. Please use discussions.

This is intentional - data is serialised your the annotation type. You can use custom serialisers the change this if you want.

**Comment by JobaDiniz** [comment]
> Gives: {'attr': {'attr1': 'Test', 'attr2': 5}}
> I would expect: {'attr': {'attr1': 'Test', 'attr2': 5, 'attr3': 'Subclass'}}

I'm sorry, but I'm facing the same issue there. Why is this not a bug? It's missing the `'attr3': 'Subclass'` attribute.

**Comment by AdeelH** [comment]
@JobaDiniz I came here with the same question today, but found the workaround in the docs here: https://docs.pydantic.dev/latest/concepts/serialization/#serializing-with-duck-typing. 

TL;DR: use `.model_dump(serialize_as_any=True)`.
